### DISCUSSION TITLE
Git Submodules manager fails to translate BitBucket Server Git URLs (SSH to HTTP) since 26.21.3

### DISCUSSION THREAD

**Discussion by ericcitaire** [renovatebot/renovate#11744] [Suggest an Idea]
### How are you running Renovate?

Self-hosted

### Please select which platform you are using if self-hosting.

Bitbucket Server

### If you're self-hosting Renovate, tell us what version of Renovate you run.

27.6.5

### Describe the bug

After upgrading from version 25 to version 27, Git Submodules manager fails to translate SSH Git URLs to HTTP Git URLs on a BitBucket Server platform.

It now translates `ssh://git@bitbucket.acme.com:2222/prj/repo.git` to `https://bitbucket.acme.com:2222/prj/repo.git` (wrong) instead of `https://bitbucket.acme.com/scm/prj/repo.git` (good).

After some digging, it seems this bug was introduced in version 26.21.3, but I don't really understand why.

### Relevant debug logs

<details><summary>Logs</summary>

```
WARN: Error mapping git submodules during extraction (repository=prj/parent)
      "err": {
        "task": {
          "commands": [
            "ls-remote",
            "--symref",
            "https://****:**redacted**@bitbucket.acme.com:2222/prj/repo.git",
            "HEAD"
          ],
          "format": "utf-8"
        },
        "message": "fatal: unable to access 'https://bitbucket.acme.com:2222/prj/repo.git/': gnutls_handshake() failed: An unexpected TLS packet was received.\n",
        "stack": "Error: fatal: unable to access 'https://bitbucket.acme.com:2222/prj/repo.git/': gnutls_handshake() failed: An unexpected TLS packet was received.\n\n    at Object.action (/usr/src/app/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:38:28)\n    at PluginStore.exec (/usr/src/app/node_modules/simple-git/src/lib/plugins/plugin-store.ts:24:29)\n    at /usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:114:40\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:111:14)\n    at GitExecutorChain.<anonymous> (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:88:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.js:5:58)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)"
      }
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by viceice** [comment]
renovate uses the repo clone urls by default, so it seems your bitbucket is behind a proxcy and you need to coinfigure the hostnames

https://github.com/renovatebot/renovate/blob/7801ae7c1652426de10557555fc7c71dbfd47f71/lib/platform/bitbucket-server/index.ts#L186-L212

**Reply by ericcitaire** [reply]
This worked fine with Renovate 25. The Bitbucket Server configuration didn't change.

**Reply by viceice** [reply]
I think you need to update the server base URL: https://confluence.atlassian.com/bitbucketserver/specify-the-bitbucket-base-url-776640392.html

**Reply by ericcitaire** [reply]
I'm not a BitBucket admin, I can't do that.
And again : this worked before with Renovate up to version 26.21.2, I've tested it. Same BitBucket, same repository every single time.

**Reply by viceice** [reply]
There was no real change in bitbucket server url handling since #8115, which was released in v24.46.0. 🤷‍♂️ 

https://github.com/renovatebot/renovate/commits/7801ae7c1652426de10557555fc7c71dbfd47f71/lib/platform/bitbucket-server/index.ts

**Reply by ericcitaire** [reply]
I really appreciate how quickly you're responding to issues and PR, etc. it's great. But this time I think you missed my point.
I'll stick to Renovate 25 for now and I'll try to find some time to fix this bug.

Cheers.

**Reply by viceice** [reply]
you can use curl `./rest/api/1.0/projects/${config.projectKey}/repos/${config.repositorySlug}` and check what's returned at `links.clone` that urls will be used by renovate as is. It prefers the `http` url

**Reply by viceice** [reply]
Above you told me it works until v26.21.2? so which is the last working version? That would help to identify the version wich breaks you.

**Reply by viceice** [reply]
we recently introduced [`gitUrl`](https://docs.renovatebot.com/self-hosted-configuration/#giturl) setting, wich can controls the used url (currently only for gitlab), maybe you would benefit from `gitUrl=endpoint` setting when implemented for bitbucket server.

**Reply by ericcitaire** [reply]
> Above you told me it works until v26.21.2? so which is the last working version? That would help to identify the version wich breaks you.

Sorry, I'm french so my english is far from perfect :). I tested a bunch of versions. v26.21.2 works fine, v26.21.3 doesn't.

**Reply by ericcitaire** [reply]
> we recently introduced [`gitUrl`](https://docs.renovatebot.com/self-hosted-configuration/#giturl) setting, wich can controls the used url (currently only for gitlab), maybe you would benefit from `gitUrl=endpoint` setting when implemented for bitbucket server.

Ok, I'll try this.

**Reply by viceice** [reply]
this won't work without a code change, so you can send a pr to support this for bitbucket server. You can checkout gitlab initRepo funtion for `gitUrl` usage.

But i think this is not your base issue, so please see my other top comment

**Reply by viceice** [reply]
> > Above you told me it works until v26.21.2? so which is the last working version? That would help to identify the version wich breaks you.
> 
> Sorry, I'm french so my english is far from perfect :). I tested a bunch of versions. v26.21.2 works fine, v26.21.3 doesn't.

I'm a german, so no english native too 😉

**Comment by viceice** [comment]
Ups, i see you have the same issue as in #11710. So maybe you should test v27.6.3, which should fix this.

https://github.com/renovatebot/renovate/releases/tag/27.6.3

**Reply by ericcitaire** [reply]
Ok, I'll try this version.
Unfortunately not now because of... Docker Hub rate limits... :(

**Reply by ericcitaire** [reply]
Ok, I tried version 27.6.3. Unfortunately, it fails for another reason (another issue introduced in https://github.com/renovatebot/renovate/commit/d6d4124b11ee66f1db019d0359cc47606f8d946d). I'll file an issue for this if I can...
So I tried to apply the fix from #11722 on version 26.21.3. It doesn't fix it.
Then I tried to revert #11615 entirely. It fixes it on versions 26.21.3 to 26.21.7.

**Reply by ericcitaire** [reply]
Can we convert this discussion back to an issue please? It seems obvious to me that a fix is needed for this.
I can propose a change to revert changes #11615 entirely, but according to @rarkins in https://github.com/renovatebot/renovate/issues/11710#issuecomment-918270402, it may cause other problems.
